### DISCUSSION TITLE
v7.0.0

### DISCUSSION THREAD

**Discussion by coolsoftwaretyler** [mobxjs/mobx-state-tree#2224] [General]
## Breaking Changes

* Validate state tree instances instead of snapshots in the `SnapshotProcessor.is` override by @airhorns in https://github.com/mobxjs/mobx-state-tree/pull/2182
* Fix typings for snapshot processor by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2198
* Improve typing of the union dispatcher function by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2203
* Ensure actions/volatile state also throw when overriding props. by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2207
* Fix tsconfig and typechecking by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2200

## Features

* Type create/validate to accept an instance type  by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2199

## Fixes

* Fix model types when extending in any way. by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2218

## Tests

* Get `bun test` working locally by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2204

### TypeScript

Most of the breaking changes in this release have to do with changes in TypeScript. So you may find that you have to make adjustments to how you're typing your models, although we hope most of these changes improve your TypeScript experience overall.

### Validating Tree Instances instead of Snapshots

Introduced in https://github.com/mobxjs/mobx-state-tree/pull/2182, when comparing for equality on whether a model `.is` something, we now look at the actual instance, rather than the snapshot itself. In the past, if you were expecting snapshot processors to make two different sides of an `.is` operation inequal, you may need to find a new way to emulate that behavior, or otherwise change your logic. We are going to take a stronger, more defined stance for when `.is` should evaluate to `true`.

This is a somewhat advanced use case, and most users should not have to make any changes.

### New Error Behavior when Overriding Props

Introduced in https://github.com/mobxjs/mobx-state-tree/pull/2207, we will now throw errors when actions and volatile state override prop names. This probably would have been a bug in your application code before, but MST will now explicitly throw when it detects this pattern. You can migrate by choosing distinct names for props, volatile state, and actions.

**Full Changelog**: https://github.com/mobxjs/mobx-state-tree/compare/v6.0.1...v7.0.0

**Comment by dbstraight** [comment]
Something seems to be messing with type checking on .create...

```ts
import { types } from "mobx-state-tree";

const Model = types.model({ a: types.number });

Model.create({ a: "hello world", b: 12323123 });
```

This doesn't complain in VS Code or `bun x tsc --noEmit`. It throws an error if you try to run it, but v6 gives a type error at compile time:

```
index.ts:5:16 - error TS2322: Type 'string' is not assignable to type 'number'.

5 Model.create({ a: "hello world", b: 12323123 });
                 ~


Found 1 error in index.ts:5
```

It is very possible the error is mine, but this is literally a brand new `bun init`, `bun add mobx mobx-state-tree` and then put the above code in index.ts.

**Reply by coolsoftwaretyler** [reply]
Hey @dbstraight - I think that's on our end. We made some changes to the create method. Thanks for the reproduction. We will investigate and figure it out soon!

Hope you're doing well. :)

**Reply by coolsoftwaretyler** [reply]
I'm going to move this discussion to an issue: https://github.com/mobxjs/mobx-state-tree/issues/2233

**Reply by coolsoftwaretyler** [reply]
Hey @dbstraight - we have some initial investigation. This might be an issue with your TS version. We're going to see if we can improve backwards compatibility, but figured you'd want to see the follow up in https://github.com/mobxjs/mobx-state-tree/issues/2233 so you can choose to upgrade TS proactively, or keep an eye on what we end up doing in case a fix is possible.
